sklearn/preprocessing/tests/test_polynomial.py

```diff
@@ -788,6 +793,207 @@ def test_polynomial_features_csr_X_dim_edges(deg, dim, interaction_only):
+        max_indices = _calc_expanded_nnz(n_features, interaction_only, deg) - 1
+        cumulative_size += max_indices + 1
+        needs_int64 = max(max_indices, max_indptr) > max_int32
+        has_bug |= not needs_int64 and cumulative_size > max_int32
+
+    if sp_version < parse_version("1.8.0") and has_bug:
+        msg = r"In scipy versions `<1.8.0`, the function `scipy.sparse.hstack`"
+        with pytest.raises(ValueError, match=msg):
+            X_trans = pf.fit_transform(X)
+        return
+
+    # When `n_features>=65535`, `scipy.sparse.hstack` may not use the right
+    # dtype for representing indices and indptr if `n_features` is still
+    # small enough so that each block matrix's indices and indptr arrays
+    # can be represnented with `np.int32`. We test `n_features==65535`
+    # since it is guaranteed to run into this bug.
+    if (
+        sp_version < parse_version("1.9.2")
+        and n_features == 65535
+        and degree == 2
+        and not interaction_only
+    ):
+        msg = r"In scipy versions `<1.9.2`, the function `scipy.sparse.hstack`"
+        with pytest.raises(ValueError, match=msg):
+            X_trans = pf.fit_transform(X)
+        return
+    X_trans = pf.fit_transform(X)
+
+    expected_dtype = np.int64 if num_combinations > np.iinfo(np.int32).max else np.int32
+    higher_order_count = (degree - 1) * int(not interaction_only)
+    expected_nnz = 1 + higher_order_count + int(include_bias)
+    assert X_trans.dtype == X.dtype
+    assert X_trans.shape == (1, pf.n_output_features_)
+    assert X_trans.indptr.dtype == X_trans.indices.dtype == expected_dtype
+    assert X_trans.nnz == expected_nnz
+    for idx in range(higher_order_count):
+        assert X_trans[0, target_indices[idx]] == pytest.approx(1.0)
+
+    offset = interaction_only * n_features
+    if degree == 3:
+        offset *= 1 + n_features
+    assert pf.n_output_features_ == target_indices[degree - 1] + 1 - offset
+
+
+@pytest.mark.parametrize("interaction_only", [True, False])
+@pytest.mark.parametrize("include_bias", [True, False])
+def test_csr_polynomial_expansion_too_large_to_index(interaction_only, include_bias):
+    n_features = np.iinfo(np.int64).max // 2
+    data = [1.0]
+    row = [0]
+    col = [n_features - 1]
+    X = sparse.csr_matrix((data, (row, col)))
+    pf = PolynomialFeatures(
+        interaction_only=interaction_only, include_bias=include_bias, degree=(2, 2)
+    )
+    msg = (
+        "The output that would result from the current configuration is too large to be"
+        " indexed"
+    )
+    with pytest.raises(ValueError, match=msg):
+        pf.fit(X)
+    with pytest.raises(ValueError, match=msg):
+        pf.fit_transform(X)
+
+
 def test_polynomial_features_behaviour_on_zero_degree():
     """Check that PolynomialFeatures raises error when degree=0 and include_bias=False,
     and output a single constant column when include_bias=True
```

Comment:
Could you also please include a test that XFAIL on Windows (with MSVC) to document a case of unmitigated integer overflow in `PolynomialFeatures.fit_transform` that would work with `int64` indexed output as long as the C/C++ compilers allows us to use `int128` typed intermediate variables?

Reply:
Note that this is different from (a complement to) what I suggest in https://github.com/scikit-learn/scikit-learn/pull/23731#discussion_r1150287129: the former would check our assumptions on the behavior of the compilers we typically use to compile scikit-learn while the latter would check the impact of this limitation to check that it can actually impact the behavior of `PolynomialFeatures` on such platforms.

Reply:
I have added `test_csr_polynomial_expansion_windows_fail` to address this